File: tests/python/gaia-ui-tests/gaiatest/apps/clock/regions/stopwatch.py
 # This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.

from marionette_driver import expected, By, Wait

from gaiatest.apps.clock.app import Clock
from gaiatest.apps.base import PageRegion



class StopWatch(Clock):
    _stopwatch_view_locator = (By.CSS_SELECTOR,'#stopwatch-panel.active')
    _stopwatch_time_locator = (By.CSS_SELECTOR, ".stopwatch-time")
    _stopwatch_start_locator = (By.CSS_SELECTOR, '#stopwatch-controls .stopwatch-start')
    _stopwatch_pause_locator = (By.CSS_SELECTOR, '#stopwatch-controls .stopwatch-pause')
    _stopwatch_reset_locator = (By.CSS_SELECTOR, '#stopwatch-controls .stopwatch-reset')
    _stopwatch_resume_locator = (By.CSS_SELECTOR, '#stopwatch-controls .stopwatch-resume')
    _stopwatch_lap_locator = (By.CSS_SELECTOR, '#stopwatch-controls .stopwatch-lap')
    _all_laps_locator = (By.CSS_SELECTOR, '.stopwatch-laps li')

    def __init__(self, marionette):
        Clock.__init__(self, marionette)
        view = self.marionette.find_element(*self._stopwatch_view_locator)
        Wait(self.marionette).until(lambda m: view.location['x'] == 0 and view.is_displayed())

    @property
    def current_time(self):
        current_time = Wait(self.marionette).until(expected.element_present(*self._stopwatch_time_locator))
        return current_time.text

    @property
    def laps(self):
        return [self.LapItem(self.marionette, lap) for lap in self.marionette.find_elements(*self._all_laps_locator)]

    def start(self):
        Wait(self.marionette).until(expected.element_present(*self._stopwatch_start_locator)).tap()

    def pause(self):
        Wait(self.marionette).until(expected.element_present(*self._stopwatch_pause_locator)).tap()

    def reset(self):
        Wait(self.marionette).until(expected.element_present(*self._stopwatch_reset_locator)).tap()

    def resume(self):
        Wait(self.marionette).until(expected.element_present(*self._stopwatch_resume_locator)).tap()

    def record_lap(self):
        Wait(self.marionette).until(expected.element_present(*self._stopwatch_lap_locator)).tap()

    class LapItem(PageRegion):
        _lap_name = (By.CSS_SELECTOR, '.lap-name')
        _lap_time = (By.CSS_SELECTOR, '.lap-duration')

        @property
        def name(self):
            return self.root_element.find_element(*self._lap_name).text

        @property
        def time(self):
            return self.root_element.find_element(*self._lap_time).text

File: apps/end2end/clock/regions/stopwatch.js
'use strict';

var PageRegion = require(
    '../../../../shared/test/integration/helpers/page_region');
var MarionetteSearchTypes = require(
    '../../../../shared/test/integration/helpers/marionette_search_types');

function Stopwatch(client) {
    PageRegion.call(this, client,
        Stopwatch.prototype.selectors._stopwatchPanelLocator);
    var viewLocator = this.findElement(
        Stopwatch.prototype.selectors._stopwatchPanelLocator);
    this.client.waitFor(function() {
            return viewLocator.displayed() && viewLocator.rect().x === 0;
        }
    );
}

Stopwatch.prototype = Object.create(PageRegion.prototype);
Stopwatch.prototype.constructor =  Stopwatch;

Stopwatch.prototype.selectors = {
    _stopwatchPanelLocator : {'by':MarionetteSearchTypes.prototype.cssId ,
        'locator': 'stopwatch-panel' },
    _stopwatchTimeLocator : {'by': MarionetteSearchTypes.prototype.cssSelector,
        'locator': '.stopwatch-time'},
    _stopwatchStartLocator: {'by': MarionetteSearchTypes.prototype.cssSelector,
        'locator': '#stopwatch-controls .stopwatch-start'},
    _stopwatchPauseLocator: {'by': MarionetteSearchTypes.prototype.cssSelector,
        'locator': '#stopwatch-controls .stopwatch-pause'},
    _stopwatchResetLocator: {'by': MarionetteSearchTypes.prototype.cssSelector,
        'locator': '#stopwatch-controls .stopwatch-reset'},
    _stopwatchResumeLocator: {'by': MarionetteSearchTypes.prototype.cssSelector,
        'locator': '#stopwatch-controls .stopwatch-resume'},
    _stopwatchLapLocator: {'by': MarionetteSearchTypes.prototype.cssSelector,
        'locator': '#stopwatch-controls .stopwatch-lap'}
};

Stopwatch.prototype.getDisplayedTime = function() {
    var currentTime = this.findElement(
        Stopwatch.prototype.selectors._stopwatchTimeLocator);
    return currentTime.text();
};

Stopwatch.prototype.tapStart = function() {
    this.tap(Stopwatch.prototype.selectors._stopwatchStartLocator);
};

Stopwatch.prototype.tapPause = function() {
    this.tap(Stopwatch.prototype.selectors._stopwatchPauseLocator);
};

Stopwatch.prototype.tapReset = function() {
    this.tap(Stopwatch.prototype.selectors._stopwatchResetLocator);
};

Stopwatch.prototype.tapResume = function() {
    this.tap(Stopwatch.prototype.selectors._stopwatchResumeLocator);
};

Stopwatch.prototype.tapLap = function() {
    this.tap(Stopwatch.prototype.selectors._stopwatchLapLocator);
};

module.exports = Stopwatch;

